File: src/main/org/nlogo/gl/view/ViewManager.java
package org.nlogo.gl.view;

import java.util.ArrayList;
import java.util.List;

import org.nlogo.window.JOGLLoadingException;
import org.nlogo.window.JOGLVersionMismatchException;
import org.nlogo.api.I18N;

public strictfp class ViewManager
    implements org.nlogo.window.GLViewManagerInterface,
    org.nlogo.window.Event.LinkChild,
    org.nlogo.window.Event.LinkParent,
    org.nlogo.window.Events.PeriodicUpdateEvent.Handler,
    org.nlogo.gl.render.GLViewSettings {

  org.nlogo.agent.World world;
  View currentView;
  ObserverView observerView;
  FullscreenView fullscreenView = null;
  View turtleView;
  final org.nlogo.window.GUIWorkspace workspace;
  private boolean fullscreen;
  public final javax.swing.JFrame appWindow;
  private final java.awt.event.KeyListener keyListener;

  private boolean paintingImmediately = false;
  private boolean framesSkipped = false;

  public ViewManager(org.nlogo.window.GUIWorkspace workspace,
                     javax.swing.JFrame appWindow,
                     java.awt.event.KeyListener keyListener) {
    this.workspace = workspace;
    world = workspace.world();
    this.appWindow = appWindow;
    this.keyListener = keyListener;
  }

  //event link component
  public Object getLinkParent() {
    return appWindow;
  }

  public void open()
      throws JOGLLoadingException {
    if (observerView != null) {
      observerView.toFront();
      observerView.updatePerspectiveLabel();
    } else {
      try {
        init();
      } catch (JOGLVersionMismatchException vex) {
        org.nlogo.swing.Utils.alert(vex.getMessage(),
            I18N.gui().get("common.buttons.continue"));
      } catch (JOGLLoadingException ex) {
        if (observerView != null) {
          observerView.dispose();
          observerView = null;
        }
        throw ex;
      }
    }
  }

  public void init()
      throws JOGLLoadingException {
    org.nlogo.gl.render.JOGLException jvmex = null;
    if (!org.nlogo.gl.render.JOGLLoader.isLoaded()) {
      try {
        ClassLoader classLoader = getClass().getClassLoader();
        org.nlogo.gl.render.JOGLLoader.load(classLoader);
      } catch (org.nlogo.gl.render.JOGLException ex) {
        if (ex.throwImmediately()) {
          throw new JOGLLoadingException(ex.getMessage(), ex.t());
        } else {
          // We should try to initialize even if we have a version mismatch,
          // so we hold off on throwing the exception until the end of the
          // method. - AZS 6/27/05
          jvmex = ex;
        }
      }
    }

    // if we have a frame already, dispose of it
    if (observerView != null) {
      observerView.dispose();
    }

    try {
      observerView = new ObserverView(this, null);
      observerView.canvas().addKeyListener(keyListener);
      currentView = observerView;
      org.nlogo.awt.Utils.moveNextTo
          (observerView, appWindow);
      currentView.updatePerspectiveLabel();
      observerView.setVisible(true);
    } catch (java.lang.UnsatisfiedLinkError e) {
      throw new JOGLLoadingException
          ("NetLogo could not load the JOGL native libraries on your computer.\n\n" +
              "Write bugs@ccl.northwestern.edu for assistance.", e);
    }

    if (jvmex != null) {
      throw new JOGLVersionMismatchException(jvmex.getMessage());
    }
  }

  public void setFullscreen(boolean fullscreen) {
    if (fullscreen == isFullscreen()) {
      return;
    }

    java.awt.GraphicsDevice gd
        = java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment()
        .getDefaultScreenDevice();

    // this is necessary in order to force PatchRenderer to make a new texture,
    // since the old one won't survive the transition to fullscreen - ST 2/9/05
    world.markPatchColorsDirty();

    if (fullscreen) {
      if (!gd.isFullScreenSupported()) {
        throw new UnsupportedOperationException
            ("This graphics environment does not support full screen mode");
      }
      currentView.setVisible(true);
      appWindow.setVisible(false);
      fullscreenView = new FullscreenView(this, currentView.renderer());
      fullscreenView.canvas().addKeyListener(keyListener);
      fullscreenView.init();
      observerView.setVisible(false);
      currentView = fullscreenView;
      this.fullscreen = true;
    } else {
      appWindow.setVisible(true);
      gd.setFullScreenWindow(null);
      observerView.setVisible(true);
      observerView.updateRenderer();
      this.fullscreen = false;
      currentView = observerView;
      fullscreenView.dispose();
    }
  }

  public boolean isFullscreen() {
    return fullscreen;
  }

  public void editFinished() {
    if (currentView != null) {
      currentView.editFinished();
    }
  }

  public boolean is3D() {
    return currentView != null;
  }

  public boolean isDead() {
    return false;
  }

  // called when user closes window, or by GUIWorkspace when new model loaded.
  public void close() {
    workspace.set2DViewEnabled(true);
    if (currentView != null) {
      currentView.dispose();
    }
    observerView = null;
    currentView = null;
  }

  private final Runnable paintRunnable =
      new Runnable() {
        public void run() {
          incrementalUpdateFromEventThread();
        }
      };

  public void incrementalUpdateFromJobThread() {
    try {
      org.nlogo.awt.Utils.invokeAndWait(paintRunnable);
    } catch (InterruptedException ex) {
      repaint();
    }
  }

  public void incrementalUpdateFromEventThread() {
    // in case we get called before init() - ST 2/18/05
    if (currentView != null) {
      workspace.updateManager().beginPainting();
      currentView.display();
      workspace.updateManager().donePainting();
      currentView.updatePerspectiveLabel();
    }

  }

  public void repaint() {
    // in case we get called before init() - ST 2/18/05
    if (currentView != null) {
      workspace.updateManager().beginPainting();
      currentView.signalViewUpdate();
      workspace.updateManager().donePainting();
      currentView.updatePerspectiveLabel();
      framesSkipped = false;
    }
  }

  private boolean antiAliasing = true;

  public void antiAliasingOn(boolean antiAliasing) {
    this.antiAliasing = antiAliasing;
    if (currentView != null) {
      world.markPatchColorsDirty();
      observerView = new ObserverView(this, currentView.renderer(), currentView.getBounds());
      currentView.dispose();
      currentView = observerView;
      currentView.setVisible(true);
    }
  }

  public boolean antiAliasingOn() {
    return antiAliasing;
  }

  private boolean wireframeOn = true;

  public boolean wireframeOn() {
    return wireframeOn;
  }

  public void wireframeOn(boolean on) {
    wireframeOn = on;
  }

  public void framesSkipped() {
    framesSkipped = true;
  }

  public void paintingImmediately(boolean paintingImmediately) {
    this.paintingImmediately = paintingImmediately;
  }

  public boolean paintingImmediately() {
    return paintingImmediately;
  }

  public void paintImmediately(boolean force) {
    if (viewIsVisible() && (framesSkipped || force)) {
      paintingImmediately(true);
      repaint();
      paintingImmediately(false);
    }
  }

  public boolean viewIsVisible() {
    return currentView.isShowing();
  }

  public java.awt.Component getExportWindowFrame() {
    return currentView;
  }

  public java.awt.image.BufferedImage exportView() {
    return currentView.exportView();
  }

  private final List<Object> linkComponents = new ArrayList<Object>();

  public void addLinkComponent(Object c) {
    linkComponents.clear();
    linkComponents.add(c);
  }

  public Object[] getLinkChildren() {
    return linkComponents.toArray();
  }

  public void handle(org.nlogo.window.Events.PeriodicUpdateEvent e) {
    if (observerView != null) {
      observerView.controlStrip().updateTicks();
    }
  }

  public boolean displayOn() {
    return workspace.displaySwitchOn();
  }

  public void displayOn(boolean displayOn) {
    workspace.displaySwitchOn(displayOn);
  }

  public org.nlogo.api.ViewSettings graphicsSettings() {
    return workspace.view;
  }

  public double mouseXCor() {
    if (currentView != null) {
      return currentView.renderer().mouseXCor();
    }
    return 0.0f;
  }

  public double mouseYCor() {
    if (currentView != null) {
      return currentView.renderer().mouseYCor();
    }
    return 0.0f;
  }

  public void resetMouseCors() {
    if (currentView != null) {
      currentView.renderer().resetMouseCors();
    }
  }

  public boolean mouseDown() {
    if (currentView != null) {
      return currentView.renderer().mouseDown();
    }
    return false;
  }

  public boolean mouseInside() {
    return currentView.renderer().mouseInside();
  }

  public void shapeChanged(org.nlogo.api.Shape shape) {
    if (currentView != null) {
      if (shape instanceof org.nlogo.shape.VectorShape) {
        currentView.invalidateTurtleShape(shape.getName());
      } else {
        currentView.invalidateLinkShape(shape.getName());
      }
      repaint();
    }
  }

  public void addCustomShapes(String filename)
      throws java.io.IOException,
      org.nlogo.shape.InvalidShapeDescriptionException {
    currentView.renderer().addCustomShapes(filename);
  }

  public void displaySwitch(boolean on) {
    observerView.controlStrip().displaySwitch().setOn(on);
  }

  public boolean displaySwitch() {
    if (observerView != null) {
      return observerView.controlStrip().displaySwitch().isSelected();
    }
    throw new IllegalStateException();
  }

  public void applyNewFontSize(int fontSize, int zoom) {
    /// I think the 3D renderer grabs it's font size
    // directly from the main view so we don't
    // need to keep track of it here.
  }

  private boolean warned = false;

  boolean warned() {
    return warned;
  }

  void warned(boolean warned) {
    this.warned = warned;
  }
}

File: netlogo-gui/src/main/gl/view/ViewManager.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.gl.view

import java.awt.event.KeyListener
import java.lang.UnsatisfiedLinkError
import javax.swing.JFrame

import org.nlogo.awt.Positioning
import org.nlogo.core.{ I18N, Shape }
import org.nlogo.gl.render.GLViewSettings
import org.nlogo.swing.Utils
import org.nlogo.window.{ GUIWorkspace, JOGLLoadingException, JOGLVersionMismatchException, TickCounterLabel }

class ViewManager(val workspace: GUIWorkspace,
                  appWindow: JFrame,
                  keyListener: KeyListener)
    extends org.nlogo.window.GLViewManagerInterface
    with org.nlogo.window.Event.LinkChild
    with org.nlogo.window.Event.LinkParent
    with org.nlogo.window.Events.PeriodicUpdateEvent.Handler
    with org.nlogo.window.LinkRoot
    with GLViewSettings {

  val world = workspace.world
  var currentView: Option[View] = None
  var observerView: Option[ObserverView] = None
  val tickCounterLabel = new TickCounterLabel(workspace.world)
  addLinkComponent(tickCounterLabel)
  private var fullscreenView: Option[FullscreenView] = None
  private var fullscreen = false

  var paintingImmediately = false
  private var _framesSkipped = false
  override def framesSkipped(): Unit = { _framesSkipped = true }

  override def getLinkParent = appWindow

  @throws(classOf[JOGLLoadingException])
  def open(): Unit = {
    observerView match {
      case Some(view) =>
        view.toFront()
        view.updatePerspectiveLabel()

      case _ =>
        try {
          init()
        } catch {
          case vex: JOGLVersionMismatchException =>
            Utils.alert(vex.getMessage, I18N.gui.get("common.buttons.continue"))

          case ex: JOGLLoadingException =>
            observerView.foreach(_.dispose())
            observerView = None

            throw ex
        }
    }

    syncTheme()
  }

  def init(): Unit = {
    // if we have a frame already, dispose of it
    observerView.foreach(_.dispose())

    try {
      val view = new ObserverView(this, null)
      observerView = Option(view)
      view.canvas.addKeyListener(keyListener)
      currentView = observerView
      Positioning.moveNextTo(view, appWindow)
      currentView.foreach(_.updatePerspectiveLabel())
      view.setVisible(true)
    } catch {
      case e: UnsatisfiedLinkError =>
        throw new JOGLLoadingException(
          "NetLogo could not load the JOGL native libraries on your computer.\n\n" +
            "Write bugs@ccl.northwestern.edu for assistance.", e)
    }
  }

  def setFullscreen(fullscreen: Boolean): Unit = {
    if (fullscreen != isFullscreen) {
      val gd = java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment.getDefaultScreenDevice
      // this is necessary in order to force PatchRenderer to make a new texture, since the old one
      // won't survive the transition to fullscreen - ST 2/9/05
      world.markPatchColorsDirty()
      if (fullscreen) {
        if (!gd.isFullScreenSupported)
          throw new UnsupportedOperationException(
            "This graphics environment does not support full screen mode")
        currentView.foreach(_.setVisible(true))
        appWindow.setVisible(false)
        val view = new FullscreenView(this, currentView.map(_.renderer).orNull)
        fullscreenView = Some(view)
        view.canvas.addKeyListener(keyListener)
        view.init()
        observerView.foreach(_.setVisible(false))
        currentView = fullscreenView
        this.fullscreen = true
      } else {
        appWindow.setVisible(true)
        gd.setFullScreenWindow(null)
        observerView.foreach(_.setVisible(true))
        observerView.foreach(_.display())
        observerView.foreach(_.updateRenderer())
        this.fullscreen = false
        currentView = observerView
        fullscreenView.foreach(_.dispose())
        fullscreenView = None
      }
    }
  }

  def isFullscreen = fullscreen

  def editFinished(): Unit = {
    currentView.foreach(_.editFinished())
  }

  def is3D = currentView.isDefined

  def isDead = false

  def close(): Unit = {
    currentView.foreach { view =>
      workspace.set2DViewEnabled(true)
      view.dispose()
      observerView = None
      currentView = None
    }
  }

  private val paintRunnable =
    new Runnable() {
      override def run(): Unit = {
        incrementalUpdateFromEventThread()
      }
    }

  def incrementalUpdateFromJobThread(): Unit = {
    try org.nlogo.awt.EventQueue.invokeAndWait(paintRunnable)
    catch {
      case ex: InterruptedException =>
        repaint()
    }
  }

  def incrementalUpdateFromEventThread(): Unit = {
    currentView.foreach { view =>
      workspace.updateManager.beginPainting()
      view.display()
      workspace.updateManager.donePainting()
      view.updatePerspectiveLabel()
    }
  }

  def repaint(): Unit = {
    currentView.foreach { view =>
      workspace.updateManager.beginPainting()
      view.display()
      workspace.updateManager.donePainting()
      view.updatePerspectiveLabel()
      _framesSkipped = false
    }
  }

  private var antiAliasing = true

  def antiAliasingOn(antiAliasing: Boolean): Unit = {
    this.antiAliasing = antiAliasing
    currentView.foreach { view =>
      world.markPatchColorsDirty()
      observerView = Option(new ObserverView(this, view.renderer, view.getBounds))
      view.dispose()
      currentView = observerView
      currentView.foreach(_.setVisible(true))
    }
  }

  def antiAliasingOn = antiAliasing

  var wireframeOn = true

  def setWireframeOn(on: Boolean): Unit = {
    wireframeOn = on
  }

  def paintImmediately(force: Boolean): Unit = {
    if (viewIsVisible && (_framesSkipped || force)) {
      paintingImmediately = true
      repaint()
      paintingImmediately = false
    }
  }

  def viewIsVisible = currentView.exists(_.isShowing)

  def getExportWindowFrame: java.awt.Component =
    currentView.orNull

  def exportView = currentView.map(_.exportView).orNull

  override def addLinkComponent(c: AnyRef): Unit = {
    linkComponents.clear()
    super.addLinkComponent(c)
  }

  def handle(e: org.nlogo.window.Events.PeriodicUpdateEvent): Unit = {
    observerView.foreach(_.controlStrip.updateTicks())
  }

  def displayOn = workspace.displaySwitchOn

  def displayOn(displayOn: Boolean): Unit = {
    workspace.displaySwitchOn(displayOn)
  }

  def graphicsSettings: org.nlogo.api.ViewSettings =
    workspace.view

  def mouseXCor =
    currentView.map(_.renderer.mouseXCor).getOrElse(0f)

  def mouseYCor =
    currentView.map(_.renderer.mouseYCor).getOrElse(0f)

  def resetMouseCors(): Unit = {
    currentView.foreach(_.renderer.resetMouseCors())
  }

  def mouseDown =
    currentView.exists(_.renderer.mouseDown)

  def mouseInside = currentView.exists(_.renderer.mouseInside)

  def shapeChanged(shape: Shape): Unit = {
    currentView.foreach { view =>
      shape match {
        case _: Shape.VectorShape =>
          view.invalidateTurtleShape(shape.name)
        case _: Shape.LinkShape =>
          view.invalidateLinkShape(shape.name)
      }
      repaint()
    }
  }

  @throws(classOf[java.io.IOException])
  @throws(classOf[org.nlogo.shape.InvalidShapeDescriptionException])
  def addCustomShapes(filename: String): Unit = {
    currentView.foreach(_.renderer.addCustomShapes(filename))
  }

  def displaySwitch(on: Boolean): Unit = {
    observerView.foreach(_.controlStrip.displaySwitch.setOn(on))
  }

  def displaySwitch =
    observerView.exists(_.controlStrip.displaySwitch.isSelected)

  // I think the 3D renderer grabs it's font size directly from the main view so we don't need to
  // keep track of it here.
  def applyNewFontSize(fontSize: Int, zoom: Int): Unit = {}

  var warned = false

  override def syncTheme(): Unit = {
    observerView.foreach(_.syncTheme())
    fullscreenView.foreach(_.syncTheme())
  }
}
